homeassistant/components/tplink/light.py

```diff
@@ -177,49 +180,59 @@ def is_on(self):
-                if (
-                    self.smartbulb.color_temp is not None
-                    and self.smartbulb.color_temp != 0
-                ):
-                    self._color_temp = kelvin_to_mired(self.smartbulb.color_temp)
+                if self._supported_features & SUPPORT_COLOR_TEMP:
+                    if (
+                        self.smartbulb.color_temp is not None
+                        and self.smartbulb.color_temp != 0
+                    ):
+                        self._color_temp = kelvin_to_mired(self.smartbulb.color_temp)
 
-            if self._supported_features & SUPPORT_COLOR:
-                hue, sat, _ = self.smartbulb.hsv
-                self._hs = (hue, sat)
+                if self._supported_features & SUPPORT_COLOR:
+                    hue, sat, _ = self.smartbulb.hsv
+                    self._hs = (hue, sat)
 
-            if self.smartbulb.has_emeter:
-                self._emeter_params[ATTR_CURRENT_POWER_W] = "{:.1f}".format(
-                    self.smartbulb.current_consumption()
-                )
-                daily_statistics = self.smartbulb.get_emeter_daily()
-                monthly_statistics = self.smartbulb.get_emeter_monthly()
-                try:
-                    self._emeter_params[ATTR_DAILY_ENERGY_KWH] = "{:.3f}".format(
-                        daily_statistics[int(time.strftime("%d"))]
-                    )
-                    self._emeter_params[ATTR_MONTHLY_ENERGY_KWH] = "{:.3f}".format(
-                        monthly_statistics[int(time.strftime("%m"))]
+                if self.smartbulb.has_emeter:
+                    self._emeter_params[ATTR_CURRENT_POWER_W] = "{:.1f}".format(
+                        self.smartbulb.current_consumption()
                     )
-                except KeyError:
-                    # device returned no daily/monthly history
-                    pass
-
-            self._available = True
-
-        except (SmartDeviceException, OSError) as ex:
+                    daily_statistics = self.smartbulb.get_emeter_daily()
+                    monthly_statistics = self.smartbulb.get_emeter_monthly()
+                    try:
+                        self._emeter_params[ATTR_DAILY_ENERGY_KWH] = "{:.3f}".format(
+                            daily_statistics[int(time.strftime("%d"))]
+                        )
+                        self._emeter_params[ATTR_MONTHLY_ENERGY_KWH] = "{:.3f}".format(
+                            monthly_statistics[int(time.strftime("%m"))]
+                        )
+                    except KeyError:
+                        # device returned no daily/monthly history
+                        pass
+
+                self._available = True
+
+            except (SmartDeviceException, OSError) as ex:
+                _LOGGER.warning(
+                    f"Retrying in {SLEEP_TIME} for {self.smartbulb.host}|{self._alias} due to: {ex}"
+                )
+                time.sleep(SLEEP_TIME)
```

Comment:
Sync sleep isn't allowed in the home assistant thread pool. It risks exhausting the pool.

Using `asyncio.sleep` would be ok, but that requires to be awaited from inside a coroutine. The sync I/O work will still need to be scheduled on the executor thread pool.

Reply:
That makes sense. Are you aware of another component / device that might have had a similar issue that I could mimic?

Reply:
After going through some of the other components I believe I understand better how this should work. I am planning on submitting an update tomorrow.

Reply:
I made the changes for the switch piece but I am still missing something and I am not sure what. When I made the update async it is just causing updates for many devices to be delayed and then I get the update taking too long warnings ([homeassistant.helpers.entity] Update of switch.laptop is taking over 10 seconds) for nearly all of my tplink devices and some of my other HA components. I reviewed the dev docs and I think I did the right thing but I can't tell what I am missing.